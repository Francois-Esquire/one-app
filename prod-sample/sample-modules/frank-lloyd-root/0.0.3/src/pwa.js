/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

/* eslint-disable global-require */

export const defaultSplashScreen = 'http://localhost:3001/static/modules/frank-lloyd-root/0.0.3/assets/pwa-icon-512px.png';
export const defaultIcon = 'http://localhost:3001/static/modules/frank-lloyd-root/0.0.3/assets/pwa-icon-192px.png';
export const appleIcon = 'http://localhost:3001/static/modules/frank-lloyd-root/0.0.3/assets/pwa-icon-180px.png';

export const scope = '/';
// eslint-disable-next-line camelcase
export const start_url = '/success';
// eslint-disable-next-line camelcase
export const theme_color = '#FDB92D';

export default {
  enabled: true,
  scope,
  manifest: {
    scope,
    start_url,
    name: 'Frank Lloyd Root',
    short_name: 'frank lloyd root',
    display: 'standalone',
    theme_color,
    background_color: '#FFF',
    icons: [
      {
        src: appleIcon,
        type: 'image/png',
        sizes: '180x180',
      },
      {
        src: defaultIcon,
        type: 'image/png',
        sizes: '192x192',
      },
      {
        src: defaultSplashScreen,
        type: 'image/png',
        sizes: '512x512',
      },
    ],
  },
};
