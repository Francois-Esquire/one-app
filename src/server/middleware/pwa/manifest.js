/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

let webmanifest = null;
let webmanifestEnabled = false;

export function resetWebmanifest() {
  webmanifest = {
    name: 'One App',
    short_name: 'One App',
  };
  return webmanifest;
}

export function getWebmanifestEnabled() {
  return webmanifestEnabled;
}

export function getWebmanifest() {
  return webmanifest;
}

export function setWebmanifest(manifest) {
  if (manifest === null) webmanifest = null;
  else Object.assign(resetWebmanifest(), manifest);
  return getWebmanifest();
}

export function configureWebmanifest({ enabled = false, manifest } = {}) {
  webmanifestEnabled = enabled;
  if (enabled === false) setWebmanifest(null);
  else setWebmanifest(manifest);
}

export function webmanifestMiddleware() {
  return function webmanifestMiddlewareHandler(req, res, next) {
    if (!webmanifestEnabled) return next();
    return res
      .type('json')
      .send(getWebmanifest());
  };
}
