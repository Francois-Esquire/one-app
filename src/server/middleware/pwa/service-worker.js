/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import fs from 'fs';

let serviceWorkerScript = null;
let serviceWorkerScope = null;
let serviceWorkerType = null;
let serviceWorkerEnabled = false;

export function getServiceWorkerEnabled() {
  return serviceWorkerEnabled;
}

export function getServiceWorkerScope() {
  return serviceWorkerScope;
}

export function getServiceWorkerType() {
  return serviceWorkerType;
}

export function getServiceWorkerScript() {
  return serviceWorkerScript;
}

export function setServiceWorkerScript(value, scope) {
  if (scope) serviceWorkerScope = scope;
  else if (value) serviceWorkerScope = '/';
  else serviceWorkerScope = null;
  serviceWorkerScript = value;
  return serviceWorkerScript;
}

export function createServiceWorkerNoopScript() {
  // this file is created during build inside lib/server/middleware/pwa
  return fs.readFileSync([__dirname, 'scripts/sw.noop.js'].join('/')).toString();
}

export function createServiceWorkerScript() {
  // this file is created during build inside lib/server/middleware/pwa
  return fs.readFileSync([__dirname, 'scripts/sw.js'].join('/')).toString();
}

export function createServiceWorkerEscapeHatchScript() {
  return 'self.unregister();';
}

export function configureServiceWorker({
  enabled, escapeHatch, noop, scope,
} = {}) {
  if (escapeHatch) {
    serviceWorkerType = 'escape-hatch';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerEscapeHatchScript(), scope);
  } else if (noop) {
    serviceWorkerType = 'noop';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerNoopScript(), scope);
  } else if (enabled) {
    serviceWorkerType = 'standard';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerScript(), scope);
  } else {
    serviceWorkerType = null;
    serviceWorkerEnabled = false;
    setServiceWorkerScript(null);
  }
}

export function serviceWorkerMiddleware() {
  return function serviceWorkerMiddlewareHandler(req, res, next) {
    if (serviceWorkerEnabled === false) return next();
    return res
      .type('js')
      .set('Service-Worker-Allowed', serviceWorkerScope)
      .set('Cache-Control', 'no-store, no-cache')
      .send(getServiceWorkerScript());
  };
}
