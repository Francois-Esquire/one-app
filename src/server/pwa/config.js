/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { routes } from './createRouter';
import { isString, isBoolean, isPlainObject } from '../utils/typeChecks';
import {
  configureServiceWorker, getServiceWorkerEnabled, getServiceWorkerScope,
} from './middleware/service-worker';

const validKeys = new Map([
  ['enabled', isBoolean],
  ['escapeHatch', isBoolean],
  ['noop', isBoolean],
  ['scope', isString],
]);

let pwaConfig = null;

export function resetPWAConfig() {
  pwaConfig = {
    enabled: false,
    escapeHatch: false,
    noop: false,
  };
  return pwaConfig;
}

export function getPWAConfig() {
  if (pwaConfig === null) return resetPWAConfig();
  return pwaConfig;
}

export function setPWAConfig(value) {
  Object.assign(resetPWAConfig(), {
    enabled: getServiceWorkerEnabled(),
    scope: getServiceWorkerScope(),
  }, value);
  return pwaConfig;
}

export function getClientPWAConfig() {
  return {
    enabled: getServiceWorkerEnabled(),
    scope: getServiceWorkerScope(),
    scriptUrl: getServiceWorkerEnabled() && [routes.prefix, routes.worker].join(''),
  };
}

export function validatePWAConfig(configToValidate) {
  if (!isPlainObject(configToValidate)) {
    console.error('invalid config given to service worker (expected "object")');
    return null;
  }

  return Object.keys(configToValidate)
    .map((key) => {
      if (validKeys.has(key)) {
        const testValueType = validKeys.get(key);

        if (testValueType(configToValidate[key])) {
          return [key, configToValidate[key]];
        }

        console.warn(
          `invalid value type given for configuration key "${key}" (expected "${testValueType.name.replace('is', '')}") - ignoring`
        );
      } else console.warn(`supplied configuration key "${key}" is not a valid property - ignoring`);

      return null;
    })
    .filter((value) => !!value)
    .reduce((map, [key, value]) => ({ ...map, [key]: value }), {});
}

export function configurePWA(config) {
  const validatedConfig = validatePWAConfig(config);

  configureServiceWorker(validatedConfig);

  return setPWAConfig(validatedConfig);
}
