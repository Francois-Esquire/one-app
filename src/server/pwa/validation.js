/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export function isString(value) {
  return typeof value === 'string';
}

export function isBoolean(value) {
  return typeof value === 'boolean';
}

export function isPlainObject(value) {
  return !!value && typeof value === 'object' && Array.isArray(value) === false;
}

export const validKeys = new Map([
  ['enabled', isBoolean],
  ['escapeHatch', isBoolean],
  ['noop', isBoolean],
  ['scope', isString],
]);

export function validatePWAConfig(configToValidate) {
  if (!isPlainObject(configToValidate)) {
    console.error('invalid config given to service worker (expected "object")');
    return null;
  }

  return Object.keys(configToValidate)
    .map((key) => {
      if (!validKeys.has(key)) {
        console.warn(`supplied configuration key "${key}" is not a valid property - ignoring`);
        return null;
      }
      
      const testValueType = validKeys.get(key);
      const configToValidateValue = configToValidate[key];
      
      if (!testValueType(configToValidateValue)) {
        console.warn(
          `invalid value type given for configuration key "${key}" (expected "${testValueType.name.replace('is', '')}") - ignoring`
        );
        return null;
      }
      
      return [key, configToValidateValue];
    })
    .filter((value) => !!value)
    .reduce((map, [key, value]) => ({ ...map, [key]: value }), {});
}
