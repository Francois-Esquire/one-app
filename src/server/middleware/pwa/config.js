/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { validatePWAConfig } from './validation';
import { configureServiceWorker } from './service-worker';

const defaultPWAConfig = {
  serviceWorker: false,
  serviceWorkerRecoveryMode: false,
  serviceWorkerType: null,
  serviceWorkerScope: null,
};

let pwaConfig = { ...defaultPWAConfig };

function resetPWAConfig() {
  pwaConfig = { ...defaultPWAConfig };
}

function setPWAConfig(newConfiguration) {
  pwaConfig = newConfiguration;
  return pwaConfig;
}

export function getClientPWAConfig() {
  const { serviceWorker, serviceWorkerRecoveryMode, serviceWorkerScope } = pwaConfig;
  return {
    serviceWorker,
    serviceWorkerRecoveryMode,
    serviceWorkerScope,
    serviceWorkerScriptUrl: serviceWorker && '/_/pwa/service-worker.js',
  };
}

export function configurePWA(config) {
  // feature flag will not allow pwa/service-worker to be configured
  // it will default to a disabled state regardless if `appConfig.pwa` was provided
  if (process.env.ONE_SERVICE_WORKER !== 'true') {
    // eslint-disable-next-line no-param-reassign
    config = null;
    resetPWAConfig();
  }

  if (!config && pwaConfig.serviceWorker) {
    // if there was a previous configuration present, we want to gracefully
    // remove any remaining instances. We currently handle this client side
    // and would only need to reset the configuration when we want to decouple.
    resetPWAConfig();
  }

  const validatedConfig = config ? validatePWAConfig(config) : {};

  let enabled = false;
  let scope = null;
  let type = null;

  // the type dictates the service worker script to be used
  if (validatedConfig.escapeHatch) type = 'escape-hatch';
  else if (validatedConfig.recoveryMode) type = 'recovery';
  else if (validatedConfig.serviceWorker) type = 'standard';

  if (type) {
    // with a valid type, we can be sure that the service worker will be
    // ready to enable and assign a scope
    enabled = true;
    scope = validatedConfig.scope || '/';
  }

  configureServiceWorker({
    type,
    scope,
  });

  return setPWAConfig({
    serviceWorker: enabled,
    serviceWorkerScope: scope,
    serviceWorkerType: type,
    serviceWorkerRecoveryMode: ['escape-hatch', 'recovery'].includes(type),
  });
}
