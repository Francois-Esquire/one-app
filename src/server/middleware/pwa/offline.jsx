/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { composeModules, RenderModule } from 'holocron';

import { getClientPWAConfig } from './config';
import {
  renderI18nScript, renderModuleScripts, serializeClientInitialState, modernBrowserChunkAssets,
} from '../sendHtml';
import renderModuleStyles from '../../utils/renderModuleStyles';
import { renderShell, renderInitialState, renderScriptTag } from '../../utils/renderShell';
import { getClientStateConfig } from '../../utils/stateConfig';
import { renderForStaticMarkup } from '../../utils/reactRendering';
import jsonStringifyForScript from '../../utils/jsonStringifyForScript';
import readJsonFile from '../../utils/readJsonFile';

const { buildVersion } = readJsonFile('../../../.build-meta.json');
const bundleIntegrity = readJsonFile('../../../bundle.integrity.manifest.json');

export default function createOfflineShellMiddleware() {
  const bundle = 'browser';
  const isDevelopment = process.env.NODE_ENV === 'development';

  return async function offlineShellMiddleware(req, res) {
    const { store, clientModuleMapCache } = req;
    const { scriptNonce } = res;
    const clientConfig = getClientStateConfig();
    const pwaMetaData = getClientPWAConfig();

    const cdnUrl = `${clientConfig.cdnUrl || '/_/static/'}app/${buildVersion}`;
    const rootModuleName = store.getState().getIn(['config', 'rootModuleName']);
    const activeLocale = store.getState().getIn(['intl', 'activeLocale']);

    const [app, appVendors, runtime, , vendors] = ['app.js'].concat(modernBrowserChunkAssets).map((resource) => ({
      src: `${cdnUrl}/${resource}`,
      integrity: bundleIntegrity[resource],
    })).map(({ src, integrity }) => renderScriptTag(src, [`integrity="${integrity}"`]));

    await store.dispatch(composeModules([{ name: rootModuleName }]));

    const { renderedString: appHtml, helmetInfo } = renderForStaticMarkup(
      <Provider store={store}>
        <RenderModule moduleName={rootModuleName} />
      </Provider>
    );

    const {
      htmlAttributes, bodyAttributes, base, title, meta, link, style, script, noscript,
    } = helmetInfo
      ? Object.entries(helmetInfo)
        .reduce((object, [key, value]) => ({
          ...object,
          [key]: value && typeof value === 'object' ? value.toString() : value || '',
        }), {})
      : {};

    const clientInitialState = store.getState();

    const styles = [].concat(renderModuleStyles(store), style).join('');
    const scripts = [].concat([
      appVendors,
      runtime,
      vendors,
      renderI18nScript(clientInitialState, cdnUrl),
      renderModuleScripts({
        clientInitialState,
        moduleMap: clientModuleMapCache[bundle],
        isDevelopmentEnv: isDevelopment,
        bundle,
      }),
      app,
    ], script).join('');
    const initialState = renderInitialState({
      config: {
        isDevelopment,
        scriptNonce,
      },
      // dictates if the client app should start rendering with
      // either ReactDOM method listed below
      __render_mode__: '\'render\'',
      // we expect the cdnUrl = 'https://example.com/cdn' || '/_/static/'
      // and we combine with the buildPath and add a trailing slash
      __webpack_public_path__: `'${cdnUrl}/'`,
      // we toggle between `browser` and `legacyBrowser` depending on `bundleType`
      __holocron_module_bundle_type__: '\'browser\'',
      // The client module map is derived from the original module map
      // and is expected to be stringified here, we do not wrap it in quotes
      // as we want it directly usable
      __CLIENT_HOLOCRON_MODULE_MAP__: jsonStringifyForScript(
        clientModuleMapCache.browser
      ),
      // The client initial state after it has serialized
      // the immutable state and stringified
      __INITIAL_STATE__: jsonStringifyForScript(
        serializeClientInitialState(clientInitialState)),
      // used to configure the client-side PWA initialization
      __pwa_metadata__: jsonStringifyForScript(pwaMetaData),
    }, {
      manifest: pwaMetaData.webManifestUrl ? [pwaMetaData.webManifestUrl] : null,
    });

    res
      .status(200)
      .type('html')
      .send(
        renderShell({
          lang: activeLocale,
          appHtml,
          scripts,
          styles,
          meta,
          base,
          title,
          links: link,
          htmlAttrs: htmlAttributes,
          bodyAttrs: bodyAttributes,
          noScripts: noscript,
          initialState,
        })
      );
  };
}
