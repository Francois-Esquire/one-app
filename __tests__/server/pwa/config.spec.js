/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import * as pwaConfig from '../../../src/server/pwa/config';

const {
  resetPWAConfig,
  getPWAConfig,
  getClientPWAConfig,
  setPWAConfig,
  validatePWAConfig,
  configurePWA,
} = pwaConfig;

jest.mock('fs', () => ({
  readFileSync: (filePath) => ({ toString: () => (filePath.endsWith('noop.js') ? '[service-worker-noop-script]' : '[service-worker-script]') }),
}));

describe('pwa configuration', () => {
  it('should export components and router', () => {
    expect(Object.entries(pwaConfig)).toMatchSnapshot();
  });

  test('getters return default state', () => {
    expect(getPWAConfig()).toMatchObject({
      enabled: false,
      escapeHatch: false,
      noop: false,
    });
    expect(getClientPWAConfig()).toMatchObject({
      enabled: false,
      scope: null,
      scriptUrl: false,
    });
  });

  test('setting service worker configuration', () => {
    expect(setPWAConfig({
      enabled: true,
    })).toMatchObject({
      enabled: true,
      escapeHatch: false,
      noop: false,
    });
    expect(setPWAConfig({
      escapeHatch: true,
    })).toMatchObject({
      enabled: false,
      escapeHatch: true,
      noop: false,
    });
    expect(setPWAConfig({
      noop: true,
    })).toMatchObject({
      enabled: false,
      escapeHatch: false,
      noop: true,
    });
  });

  test('resetPWAConfig', () => {
    expect(resetPWAConfig()).toEqual({
      enabled: false,
      escapeHatch: false,
      noop: false,
    });
    expect(getPWAConfig()).toEqual({
      enabled: false,
      escapeHatch: false,
      noop: false,
    });
  });

  describe('validation', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    beforeAll(() => {
      jest.spyOn(console, 'warn');
      jest.spyOn(console, 'error');
      console.warn.mockImplementation();
      console.error.mockImplementation();
    });

    test('invalid configuration object informs the user', () => {
      expect(validatePWAConfig(null)).toEqual(null);
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error).toHaveBeenCalledWith('invalid config given to service worker (expected "object")');
      expect(validatePWAConfig([])).toEqual(null);
      expect(validatePWAConfig(true)).toEqual(null);
    });

    test('invalid configuration keys informs the user and ignores them', () => {
      expect(validatePWAConfig({
        random: 'key',
        foo: 'bar',
      })).toEqual({});
      expect(console.warn).toHaveBeenCalledTimes(2);
      expect(console.warn).toHaveBeenCalledWith('supplied configuration key "random" is not a valid property - ignoring');
      expect(console.warn).toHaveBeenCalledWith('supplied configuration key "foo" is not a valid property - ignoring');
    });

    test('invalid configuration values for keys informs the user and ignores them', () => {
      expect(validatePWAConfig({
        enabled: 'true',
        scope: 42,
      })).toEqual({});
      expect(console.warn).toHaveBeenCalledTimes(2);
      expect(console.warn).toHaveBeenCalledWith('invalid value type given for configuration key "enabled" (expected "Boolean") - ignoring');
      expect(console.warn).toHaveBeenCalledWith('invalid value type given for configuration key "scope" (expected "String") - ignoring');
    });

    test('valid keys emits no warnings or errors and returns valid configuration', () => {
      const validConfig = {
        enabled: true,
        scope: '/',
      };

      expect(validatePWAConfig(validConfig)).toEqual(validConfig);
      expect(console.warn).not.toHaveBeenCalled();
      expect(console.error).not.toHaveBeenCalled();
    });
  });

  describe('configuration', () => {
    test('configuring the service worker and webmanifest to be enabled', () => {
      const config = {
        enabled: true,
      };
      const expectedConfig = {
        enabled: true,
        escapeHatch: false,
        noop: false,
        scope: '/',
      };

      expect(configurePWA(config)).toMatchObject(expectedConfig);
      expect(getPWAConfig()).toMatchObject(expectedConfig);
      expect(getClientPWAConfig()).toMatchObject({
        enabled: true,
        scope: '/',
        scriptUrl: expect.any(String),
      });
    });

    test('disabling PWA configuration', () => {
      expect(configurePWA({ enabled: false })).toMatchObject({ enabled: false });
      expect(getPWAConfig()).toMatchObject({ enabled: false });
      expect(getClientPWAConfig()).toMatchObject({
        enabled: false, scope: null, scriptUrl: false,
      });
    });
  });
});
