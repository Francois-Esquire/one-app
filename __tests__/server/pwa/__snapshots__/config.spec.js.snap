// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pwa configuration should export components and router 1`] = `
Array [
  Array [
    "resetPWAConfig",
    [Function],
  ],
  Array [
    "getPWAConfig",
    [Function],
  ],
  Array [
    "setPWAConfig",
    [Function],
  ],
  Array [
    "getClientPWAConfig",
    [Function],
  ],
  Array [
    "validatePWAConfig",
    [Function],
  ],
  Array [
    "configurePWA",
    [Function],
  ],
]
`;
